src/tools/illink/src/linker/Linker.Steps/MarkStep.MethodDefinitionDependencyNode.cs

```diff
@@ -0,0 +1,45 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using ILCompiler.DependencyAnalysisFramework;
+using Mono.Cecil;
+
+namespace Mono.Linker.Steps
+{
+	public partial class MarkStep
+	{
+		internal sealed class MethodDefinitionDependencyNode : DependencyNodeCore<MarkStepNodeFactory>
+		{
+			readonly MethodDefinition method;
+			readonly MessageOrigin origin;
+			readonly DependencyInfo reason;
```

Comment:
I assume the medium term goal would be to get rid of these, right? We'll want to have a 1:1 relationship between methods and their nodes. A new node for different origin probably won't work for very long.

We might have different nodes for different "shades" of a method (for example, "method body" might be different from "method" - since "method" without "method body" would allow us to stub out the body).

Reply:
Yeah, I can actually fully remove those from the TypeDefinitionNode, and remove MessageOrigin from MethodDefintion in this PR without too much work. The method processing uses DependencyInfo a little bit more throughout, so I'll do that in a follow-up.

The factory uses the MethodDefinition as the key to the cache, and only creates a single node per MethodDefinition, always with the DependencyInfo from the first marking. This matches the previous behavior (if a method was processed already, we wouldn't process again with a new reason), which makes me think that the DependencyInfo and MessageOrigin in ProcessMethod never did anything interesting.

Reply:
Agreed, I have a follow up almost ready that removes the MessageOrigin and DependencyInfo from the nodes and ProcessMethod/Type methods.